### PR TITLE
backport utf8 library from 5.3

### PR DESCRIPTION
**Pull request by fperrad**
initially submitted as https://github.com/LuaJIT/LuaJIT/pull/270 and rejected.

### CODE REVIEW

**Comment by fperrad** [comment]
handle variant GC64

**Comment by siddhesh** [comment]
Do you generate utf8.h in two parts, i.e. once for LJ_GC64 and once without and then build it?  This looks OK overall, but I suppose we need some mechanism to make sure that utf8.h gets updated when utf8.lua is.

Thanks to this I noticed Makefile.dep also needs to be manually updated (i.e. the `make depend` target doesn't get run automatically).  I'll pull this patch in as is and do something about these dependency issues later.

File: src/lib_utf8.c [hunk 1]
```diff
@@ -0,0 +1,25 @@
...
+#include "lua.h"
+#include "lauxlib.h"
+#include "lualib.h"
+
+#include "lj_arch.h"
+#include "utf8.h"
+
+/* ------------------------------------------------------------------------ */
+
+LUALIB_API int luaopen_utf8(lua_State *L)
```
Comment (siddhesh) [inline src/lib_utf8.c:17]: This needs a mention in the documentation, i.e. not just the function but also the fact that we have the UTF8 library now.

File: test/lib/contents.lua [hunk 2]
```diff
@@ -148,6 +148,10 @@ do --- package.loaded
   check(loaded, "_G:coroutine:debug:io:math:os:package:string:table:thread.exdata", "bit:bit32:common:ffi:jit:table.new")
 end
 
+do --- utf8
+  check(utf8, "char:charpattern:codepoint:codes:len:offset")
```
Comment (siddhesh) [inline test/lib/contents.lua:152]: It would be nice to have some smoke tests that actually exercise the functions.

**Review by siddhesh — CHANGES_REQUESTED** [review]
A more thorough review this time:  Please re-post with the requested changes, which is basically documentation and more tests.

**Comment by fperrad** [comment]
`utf8.h` must be generated in 2 parts, each one with a specific executable `luajit` (built with and without `-DLUAJIT_ENABLE_GC64`), so the target `utf8bc` cannot fully automatized this.

`host/buildvm_libbc.h` and the target `libbc` have the same limitations.

**Comment by siddhesh** [comment]
> `utf8.h` must be generated in 2 parts, each one with a specific executable `luajit` (built with and without `-DLUAJIT_ENABLE_GC64`), so the target `utf8bc` cannot fully automatized this.

/me nods.  I was suggesting auto-generating every time, so that we don't have to keep track.  In any case, it's an orthogonal issue.  This PR itself only needs some documentation of the fact that we have utf8 library support and it's good to go.

EDIT: documentation and smoke tests :)

**Comment by fperrad** [comment]
doc and smoke tests added

5.4 version will come in another PR

**Review by siddhesh — APPROVED** [review]

**Comment by siddhesh** [comment]
Thanks @fperrad

**Pull request merged**